File: Stack/496.Next-Greater-Element-I/496.Next-Greater-Element-I.ts
function nextGreaterElement(nums1: number[], nums2: number[]): number[] {
  const ret = new Array(nums1.length).fill(-1);
  const stack = [];
  const notebook = new Map<number, number>();
  for (let i = 0; i < nums1.length; i++) {
    notebook.set(nums1[i], i);
  }

  for (let i = 0; i < nums2.length; i++) {
    if (stack.length === 0) {
      stack.push(nums2[i]);
    } else if (nums2[i] <= stack[stack.length - 1]) {
      stack.push(nums2[i]);
    } else {
      while (stack.length !== 0 && nums2[i] > stack[stack.length - 1]) {
        ret[notebook.get(stack[stack.length - 1])] = nums2[i];
        stack.pop();
      }
      stack.push(nums2[i]);
    }
  }
  return ret;
}

File: Stack/496.Next-Greater-Element-I/496.Next-Greater-Element-I.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        result = [-1] * len(nums1)

        notebook = {}
        for index, value in enumerate(nums1):
            notebook[value] = index

        stack = [nums2[0]]
        for num in nums2[1:]:
            while len(stack) and num > stack[-1]:
                lastNumber = stack.pop()
                if lastNumber in notebook:
                    result[notebook[lastNumber]] = num
            stack.append(num)
        return result
